src/Compiler/Service/ServiceAnalysis.fsi

```diff
@@ -31,3 +32,12 @@ module public UnusedDeclarations =
 
     /// Get all unused declarations in a file
     val getUnusedDeclarations: checkFileResults: FSharpCheckFileResults * isScriptFile: bool -> Async<seq<range>>
+
+module public UnnecessaryParentheses =
+
+    /// Gets the ranges of all unnecessary pairs of parentheses in a file.
+    ///
+    /// Note that this may include pairs of nested ranges each of whose
+    /// lack of necessity depends on the other's presence, such
+    /// that it is valid to remove either set of parentheses but not both.
```

Comment:
Do you think some tests could be also added on the compiler service level?

Also, it would be helpful to specify some examples in the docs here. E.g. I logically understand the above sentence but it took me a while to come up with an example.

Reply:
Yes, I'll do both.

Reply:
a7cca0f

Reply:
I added some very basic tests that directly call `FSharp.Compiler.EditorServices.UnnecessaryParentheses.getUnnecessaryParentheses` here:

a2b70f9

Is that what you were thinking? Or should they be more comprehensive, like the code fix tests?

Also, `getUnnecessaryParentheses` will actually _not_ currently return two ranges for the example that I added to the doc comment here in a7cca0f (`(x.M(y)).N`), even though it theoretically could (and did at some point).

One possibility might simply be to update `getUnnecessaryParentheses` so that it has an opinion on all known mutually-dependent nested parentheses scenarios (`(x.M(y)).N`, `id (-(-x))`, …?), instead of filtering them out at the fix-all stage. The only problem with that approach is that I don't know how we can be 100% sure that we've thought of all such possible scenarios.

If we could somehow be confident in that, or confident enough, then that would be nicer, since as it is you must explicitly apply successive rounds of the code fix to strip nested unnecessary parentheses: `((3))` → `(3)` → `3`, `((x + y) + z)` → `(x + y) + z` → `x + y + z`, etc.

Reply:
I think the new tests are enough. Since this is a complex matter, I am quite sure we might have not taken some corner cases into account - but that's fine, that's our approach, IMO as long as there are easy ways to repro and test things, we can quickly iterate and followup.

For about the same reason, I am also inclined to have this opinionated decisions about mutually dependent nested parentheses. And well, since this is a minority of cases, we can just pick the first possible option and see this in action.

Reply:
Hmm. It looks like even if we did take a stance on `(x.M(y)).N → (x.M y).N ↮ x.M(y).N`, etc., the way the fixes are currently applied during a fix-all application doesn't make sense when there are nested fixes, even when they are _not_ mutually-dependent. That's because the ranges to which any inner fixes would apply can be shifted somewhat unpredictably by the application of outer fixes: e.g., when an open paren is removed, a leading space may or may not be inserted to replace it depending on the context, etc.

The C#/VB code fix _does_ appear to handle this, though I have not yet looked at how it does so.

Reply:
...It looks like the C#/VB fix-all implementation has access to some more sophisticated AST-editing machinery than we currently do:

https://github.com/dotnet/roslyn/blob/79d1884dcdc0fb3f3c0763e372e2e010dbbcfa4a/src/Analyzers/Core/CodeFixes/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesCodeFixProvider.cs#L34-L47

I can think of a few things to experiment with, but I won't have time till next week, so maybe it can be a followup.

Reply:
Yeah I think we can take inspiration from there - thanks for looking into this.